### ISSUE TITLE
`Detections.empty()` invalidates detections, causes crashes when `Detections.merge()` is called.

### ISSUE THREAD

**Issue by LinasKo** [roboflow/supervision#943]
### Search before asking

- [X] I have searched the Supervision [issues](https://github.com/roboflow/supervision/issues) and found no similar bug report.

### Bug

This is a more general version of #928. I believe it is important enough to deserve a separate issue.

Proof found in [this Colab](https://colab.research.google.com/drive/1ktj_CIlM9mcmboo8LJUCsYgKlaPN8bJM?usp=sharing).

### Explanation
`Detections` contains these variables:
```python
xyxy: np.ndarray
mask: Optional[np.ndarray] = None
confidence: Optional[np.ndarray] = None
class_id: Optional[np.ndarray] = None
tracker_id: Optional[np.ndarray] = None
data: Dict[str, Union[np.ndarray, List]] = field(default_factory=dict)
```

Suppose we call `Detections.merge([detections_1, detections_2])`. An error is rightfully raised when the same variable is defined differently - it is `None` in `detections_1` and `np.array([])` in `detections_2`. It makes sense as we don't want to merge detections where, for example "no mask was detected" and "masks should never be detected" - that's the meaning of "None."

However, when calling `Detections.empty()`, frequently used to initialize the no-detections case, it sets a selective subset of the fields - only the `xyxy`, `condifence` and `class_id` are set.
Many models use other fields, some use trackers. If you merge with an empty detections set at some point, you're very likely to raise the aforementioned error.

Ultimately, `Detections.empty()` needs to distinguish between "fields that shouldn't exist" and "fields that are empty" and adapt that based on the type of model used / operations applied (e.g. tracking).

### Minimal Reproducible Example

Again, the Colab [the Colab](https://colab.research.google.com/drive/1ktj_CIlM9mcmboo8LJUCsYgKlaPN8bJM?usp=sharing) shows examples.

### Additional

I can help submit a PR, but this is a non-trivial bug, affecting at least 6 files, with effects on others.

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by LinasKo** [comment]
# Affected Code:
The good news is - with the current feature set, it mostly happens when `Detections.merge` is called on results from _unusual models_  - where `confidence` is missing, where `data` is set, where there's a `mask` variable but no detections.

Affected functions:
* `supervision/detection/core.py -> from_inference()`
* `supervision/detection/core.py -> from_roboflow()`
* `dataset/formats/yolo.py -> yolo_annotations_to_detections()`
* `dataset/formats/yolo.py -> load_yolo_annotations()`
* `dataset/formats/pascal_voc.py -> detections_from_xml_obj()`
* `dataset/formats/coco.py -> coco_annotations_to_detections()`
* `supervision/detection/core.py -> merge()`
* `tracker/byte_tracker/core.py -> update_with_detections()`

See [Colab](https://colab.research.google.com/drive/1ktj_CIlM9mcmboo8LJUCsYgKlaPN8bJM?usp=sharing) for more details.

**Comment by LinasKo** [comment]
# Affected code
The good news is - the error mostly appears when `Detections.merge` is called on results from _unusual models_ - where `confidence` is missing, where `data` is set, where `mask` exists (and there's no detections at least once).

Affected functions:
* `supervision/detection/core.py -> from_inference()`
* `supervision/detection/core.py -> from_roboflow()`
* `dataset/formats/yolo.py -> yolo_annotations_to_detections()`
* `dataset/formats/yolo.py -> load_yolo_annotations()`
* `dataset/formats/pascal_voc.py -> detections_from_xml_obj()`
* `dataset/formats/coco.py -> coco_annotations_to_detections()`
* `supervision/detection/core.py -> merge()`
* `tracker/byte_tracker/core.py -> update_with_detections()`

See [Colab](https://colab.research.google.com/drive/1ktj_CIlM9mcmboo8LJUCsYgKlaPN8bJM?usp=sharing) for more details.

**Comment by LinasKo** [comment]
# Mitigation Options
**TL;DR: I think Option 3 is the least-bad solution. Adds complexity in exactly one location, keeps design and architecture intact, stops errors, prevents future worries. I also advocate for two eventual changes regardless of option taken (final section)**

Here's how I see various mitigation options and their downsides. I can only speculate about the design goals and future plans of Roboflow, and I do not have access to the internal message board. So my goal now is to bring other devs up to speed and help think through the solutions.

### My Assumptions
1. `Detections` is meant to be universal and easily-understood, fitting both detection and segmentation models.
2. Therefore, it must hold undefined variables, ready to take values from different model types.
3. There's a need for `Detections.empty`, which represents empty detections of ANY type.

### Caveats
1. I have done little research on the `data` field, but found an example where it fails for the same reason.
2. I may make generalizations for which models are used less frequently than others. This comes from observing their presence in the code, and how Roboflow prioritizes them on the site.

### Option 1: Do nothing
The error has two stages:
1. Calling `Detections.empty` from specific model loaders sets it up - this happens frequently.
2. Calling `Detections.merge` triggers it - this happens rarely.

`merge` is:
* Used in `InferenceSlicer`
* Used in smoother (only partially patched in #928.)
* Exposed externally for users

Maybe it's not worth fixing? There haven't been many complaints.

The devs just need to be extra careful of using `merge` in anything new, largely relating to instance segmentation. Who knows - maybe there's not that many features left unimplemented.

### Option 2: Forbid everything affected.
(The worst approach, in my opinion)
Very radical - comment out merge and inference slicer, disallow smoothing. Who cares if there's incompatible objects around - they're not interacting with each other anyway.
(I care).

### Option 3: Change merge.
I like how `merge` works, given the assumptions. But maybe it's harmful, even if correct?

If (`xyxy`, `confidence`, `class_id`) merges with (`xyxy`, `mask`, `class_id`), maybe find a way to let it happen? Form (`xyxy`, `mask`, `confidence`, `class_id`) somehow.

(Brainstorming) When we need new values for:
* `xyxy` - make a bounding box around `mask`, 0-sized box otherwise.
* `confidence` - default to `1`.
* `class_id` - set to `-1` (raaaarely seen - e.g. `from_sam`)
* `tracker_id` - won't be an issue after #928 - in other cases `update_with_detections` ensures the ID is set.
* `mask` - expand `xyxy` into a rectangular mask.
* `data` - I don't know enough to say.
It seems to avoid both errors and architectural shifts, redefining what "compatible Detections" mean, at least until the next feature comparing multiple detections is added. Is this the least-bad solution?

### Option 4: Explicitly define what's being initialized.
`empty` sets (`xyxy`, `confidence`, `class_id`). What is that? A detection model.

Let's make it initialize (`xyxy`, `mask`, `class_id`) for segmentation models. Some won't conform - SAM returns no `class_id`. So force it, make a mock id! This way you'd have two distinct model types.

Then, either pass an egument to `Detections.empty` to tell which one you're initializing, or define two different `empty` functions. For the outliers - full field set detections and those models with `data` - either set the fields explicitly, or add a third, more flexible `Detections.empty`. Same thing.

Did you spot an issue?

Suppose you have an intermediary function. A call chain of:
`Smoother` -> `some_func` -> `specific_empty`.

Which type of `empty` should it call? Somehow, with an argument or context of some sort, `some_func` needs to know which type of variables to initialize / which `empty` to call. At present, if there's no detections such an intermediary would only receive an empty list - `[]`. No type information!

Here's a thought experiment: try inlining `Detections.empty`. I mean, replace it with the `Detections` constructor and then explicitly set variables inside. In each case, can you tell which ones should be set? Try this inside `Detections.merge` to see the issue.

### Option 4b: Frickin' Templates
We're passing the type information with function calls, effectively defining polymorphic functions. `Detections.empty<Det>()`, `Detections.empty<Seg>()` might be completely normal in other languages.

[Type parameter syntax](https://docs.python.org/3.12/library/typing.html#generics)is only introduced in 3.12. Perhaps [TypeVar](https://docs.python.org/3.8/library/typing.html), could help, but generally this would make the code less accessible to new devs.
### Option 5: Multiple Classes
Did you notice that we're approaching a multi-class solution? My claim is that `Detections` can be at least two distinct objects - the detection and the segmentation result.

An option is to change the core design, split them in two, then define strict non-nullable members, adjust everything else to match that.

`Detections` becomes a base class passed to users and annotators, those have to frequently check contents. Notice that with nullable members similar checks might already be happening (I haven't checked).

Downside:
* Lots of changes
* Less flexibility - it defeats the purpose, if we made the types in separate classes nullable.
* Edge cases - `data` and non-standard parameter sets are an issue - this solution is an issue.

### Option 6: Get rid of nullable types
This is my least thought-out option. Null values serve a purpose, representing invalid members - E.g. "This `Detections` result will never have `confidence`". This requires extra validation, disallows safely having a strong `Detections.empty` function such as the one we have now.

Perhaps `Optional` status can be revoked. There needs to be logic such that whether there's an empty `mask` or never any mask would look the same. Is there harm to that?

I'm think so. Null checks can be removed in places like `__iter__()`, but in a `for` loop, is there a way for users to easily tell which fields might suddenly produce values?

### Essential Changes
I advocate for these two changes to be made _eventually._
1. `Detections.empty` is meant to construct empty detections for ANY model. Therefore it HAS to be weaker-defined than any model. The code now sets `(xyxy, confidence, class_id)`. Does every model have at least all of these? No. Trim it down until it can be used as the constructor for `Detections` or you'll be setting fields that normal model responses won't ever have and constructing invalid objects.
   _After this is done - can `Detections` constructor be used instead of `Detections.empty()`?_
2. Visually separate `tracker_id` from other members. Is is addressed differently, has a special set of methods ensuring its existence and is not produced by the models. A single empty line would give a hint to maintainers of its special status - miniscule thing, but it would help learn the code faster.

**Comment by LinasKo** [comment]
We've now implemented option 3 in https://github.com/roboflow/supervision/pull/1177

Ideally, I'd like to see **an attempt at** trimming down `Detections.empty` to just `xyxy` before closing this.

Would that cause issues? Would that be easy fine? Would that entail too much work to be worth it? I'd like to know that.

**Comment by Borda** [comment]
Hi @LinasKo — the core merge crash was fixed via #1177 (shipped in 0.21.0). The remaining work — trimming `Detections.empty()` to just `xyxy` — is more of a cleanup/enhancement than a bug at this point. Closing this issue. If the `empty()` simplification is still desired, a new enhancement issue would be the right place to track it. Please reopen if the merge crash itself resurfaces.
